src/game/UIExitMenu.cpp

```diff
@@ -0,0 +1,180 @@
+// UIChat.cpp: implementation of the CUIChat class.
+//
+//////////////////////////////////////////////////////////////////////
+
+#include "StdAfx.h"
+#include "UIExitMenu.h"
+#include "PlayerMySelf.h"
+
+#include "APISocket.h"
+#include "PacketDef.h"
+#include "GameProcMain.h"
+#include "GameProcedure.h"
+#include "GameProcLogin.h"
+#include "GameDef.h"
+
+#include "UIMessageWnd.h"
+#include "UIManager.h"
+#include "UIPartyOrForce.h"
+#include "SubProcPerTrade.h"
+
+#include "Resource.h"
+
+//////////////////////////////////////////////////////////////////////
+// Construction/Destruction
+//////////////////////////////////////////////////////////////////////
+
+CUIExitMenu::CUIExitMenu() {
+    m_pBtn_chr = NULL;
+    m_pBtn_exit = NULL;
+    m_pBtn_option = NULL;
+    m_pBtn_cancel = NULL;
+}
+
+CUIExitMenu::~CUIExitMenu() {}
+
+void CUIExitMenu::Release() {
+    CN3UIBase::Release();
+}
+
+void CUIExitMenu::SelectCharacter() {
+
+    /*
+	
+  v0 = s_pPlayer;
+    if (s_pPlayer && s_pFX) {
+        s_pPlayer->m_bMoveUnk = 0;
+        s_pPlayer->field_9EC = 0;
+        CN3FXMgr::TriggerCursor((_DWORD *)s_pFX, s_pPlayer->m_InfoBase_iID, s_pPlayer->m_InfoBase_iID,
+                                30001, // fx/target_pointer.fxb
+                                30001, FX_BUNDLE_MOVE_DIR_FLEXABLETARGET);
+        v0 = s_pPlayer;
+    }
+    
+    */
+    // TODO: CN3FXMgr::TriggerCursor
+    CGameBase::s_pPlayer->StunRelease();
+
+    if (CGameProcedure::s_pProcMain->m_pSubProcPerTrade) {
+
+        // TODO: MERCHANT CLEAR MODEL
+
+        CGameProcedure::s_pProcMain->m_pSubProcPerTrade->LeavePerTradeState(PER_TRADE_RESULT_MY_CANCEL);
+    }
+
+    if (CGameProcedure::s_pProcMain->m_pUIPartyOrForce) {
+        CGameProcedure::s_pProcMain->m_pUIPartyOrForce->MemberDestroy();
+    }
+
+    std::string szIP = CGameProcedure::s_pSocket->GetCurrentIP();
+    DWORD       dwPort = CGameProcedure::s_pSocket->GetCurrentPort();
+
+    CGameProcedure::s_bNeedReportConnectionClosed = false;
+    CGameProcedure::s_pSocket->Disconnect();
+    Sleep(4000);
+    CGameProcedure::s_pSocket->Connect(CGameProcedure::s_hWndBase, szIP.c_str(), dwPort);
+    CGameProcedure::s_bNeedReportConnectionClosed = true;
+
+    CGameProcedure::MsgSend_GameServerLogIn();
+    CGameProcedure::ProcActiveSet((CGameProcedure *)CGameProcedure::s_pProcCharacterSelect);
+}
+
+void CUIExitMenu::AddWarningMessage(DWORD idMessage, bool visbleMenu) {
+
+    if (!CGameProcedure::s_pProcMain->m_pUIChatDlg) {
+        return;
+    }
+
+    std::string szMsg;
+    ::_LoadStringFromResource(idMessage, szMsg);
+
+    CGameProcedure::s_pProcMain->m_pUIChatDlg->AddChatMsg(N3_CHAT_NORMAL, szMsg, 0xFFFF0000);
+
+    SetVisible(visbleMenu);
+}
```

Comment:
What you think to move this function to GameProcMain ? In future if i wanna change something it will be easy to change and it's less code instead copy this code in four places :)

Reply:
Well, we could theoretically move it to `CGameProcMain`, or actually move it inside `CUIChat` sounds more appropriate, BUT! I generally think this function is limiting the caller. Surely it reduces code duplication in this specific scenario, but it's again very limited and not well defined. Let me explain why:
When we call `_LoadStringFromResource`, we retrieve a string from `Texts.tbl`. This string may be just a regular string or a string with format specifiers. Example would be "You have %d hp.". In this specific case, you assume that it would just be a regular string. Now if I want to add a chat warning and actually format it with the hp value, I simply can't, unless again you add another argument to this function to receive a value. What if the string you're loading by id format specifier is expecting a `string`, `float`, `int`, `double` etc...you'll will need to start overly complicate things that should be trivial with overloaded functions, or even worse for this specific scenario; [template specialization](https://en.cppreference.com/w/cpp/language/template_specialization).
Oh, and also that `SetVisible` call at the end, you'll need to pass another argument to this function to receive a `CN3UIBase` pointer.
The `AddChatMsg` call there is also quite limiting, because it hardcodes `N3_CHAT_NORMAL`, whilst you could have multiple chat modes. Also the hardcoded `0xFFFF0000` yellow color for the chat text.
I can continue, but I think you get the gist of my point here.

I think what you want here instead of a function, is a macro within the scope of `CUIExitMenu::ReceiveMessage`. That sounds reasonable to me.

Reply:
Done here: https://github.com/ko4life-net/ko/pull/166/commits/6f6417edb75a3217660987209969d36c1a0e94c0